### ISSUE TITLE
IR: port optimizations done in the wasm compiler to IR post-processing

### ISSUE THREAD

**Issue by srenatus** [open-policy-agent/opa#5185]
There are a few optimizations in the wasm compiler. For example, if it encounters a `ir.NotEqualStmt` with two constants, it will
- emit nothing if they are equal
- emit a break if they are not

Effectively avoiding the need to do the same thing at runtime.
```go
		case *ir.NotEqualStmt:
			if stmt.A == stmt.B { // same local, same bool constant, or same string constant
				instrs = append(instrs, instruction.Br{Index: 0})
				continue
			}
			_, okA := stmt.A.Value.(ir.Bool)
			if _, okB := stmt.B.Value.(ir.Bool); okA && okB {
				// not equal (checked above), but both booleans => not equal
				continue
			}
			_, okA = stmt.A.Value.(ir.StringIndex)
			if _, okB := stmt.B.Value.(ir.StringIndex); okA && okB {
				// not equal (checked above), but both strings => not equal
				continue
			}
			instrs = append(instrs, c.instrRead(stmt.A))
			instrs = append(instrs, c.instrRead(stmt.B))
			instrs = append(instrs, instruction.Call{Index: c.function(opaValueCompare)})
			instrs = append(instrs, instruction.I32Eqz{})
			instrs = append(instrs, instruction.BrIf{Index: 0})
```

👉 https://sourcegraph.com/github.com/open-policy-agent/opa@ab385e9c172f68c42bc3924dcbcad29c4d937407/-/blob/internal/compiler/wasm/wasm.go?L1095-1114

Similar things happen with `EqualStmt` and some others, like `IsObjectStmt`.

Since we now have consumers of the IR, it would make sense to do investigate if these optimizations could be done on the IR alone.

**Comment by srenatus** [comment]
There's more we could do -- "assign trains" like these could be combined:
```
| | | | *ir.AssignVarStmt &{Source:{Value:Local<4>} Target:Local<5> Location:{File:0 Col:25 Row:5 file:module-0.rego text:f0}}
| | | | *ir.AssignVarStmt &{Source:{Value:Local<5>} Target:Local<7> Location:{File:0 Col:7 Row:5 file:module-0.rego text:f(true)}}
| | | | *ir.AssignVarStmt &{Source:{Value:Local<7>} Target:Local<8> Location:{File:0 Col:3 Row:5 file:module-0.rego text:y = f(true) with f as f0}}
```
Either assign it straight to `Local<8>`, or save locals and replace all usage of `Local<7>` and `Local<8>` by `Local<5>`.

**Comment by srenatus** [comment]
Fixes by #6386.
